### PR TITLE
Add database router to fix false positives on relation mismatch.

### LINKED ISSUE #10467: giftCardCreate mutation tries to create GiftCard and GiftCardTag on different databases
[Error from Sentry](https://sentry.io/organizations/saleor/issues/3526888361/):
Cannot add "<GiftCardTag: GiftCardTag object (290)>": instance is on database "replica", value is on database "default"

This happened on staging when executing giftCardCreate mutation. Apparently GiftCard instance was created on `replica` database and the Tag is created on `default` db.

### PR DESCRIPTION
**Pull request by dekoza**
I want to merge this change because...

Fixes #10467 

# Pull Request Checklist

* [x] Privileged queries and mutations are guarded by proper permission checks
* [x] Database queries are optimized and the number of queries is constant
* [x] Database migration files are up to date
* [x] The changes are tested
* [x] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

Internal change, no need to update changelog.

### CODE REVIEW

File: saleor/core/db_routers.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+from django.conf import settings
+
+
+class PrimaryReplicaRouter:
+    def db_for_read(self, model, **hints):
+        """Read from replica."""
+        return settings.DATABASE_CONNECTION_REPLICA_NAME
```
Comment (fowczarek) [inline saleor/core/db_routers.py:7]: Won't this switch our whole read traffic to replicas? 
In my opinion, we shouldn't switch all read traffic. If we want to read from a replica we should do it explicitly. 
When we switch to always read from replica we could cause that situation when we respond to mutation from replica before this database receives propagated data from primary database.

Comment (patrys) [inline saleor/core/db_routers.py:7]: We can't implement this method. We don't want to read from a replica during mutations, especially not inside transactions when we do `SELECT ... FOR UPDATE`.

File: saleor/core/db_routers.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    def db_for_write(self, model, **hints):
+        """Write only to primary."""
+        return settings.DATABASE_CONNECTION_DEFAULT_NAME
+
+    def allow_relation(self, obj1, obj2, **hints):
+        """All relations are allowed as we don't have pool separation."""
+        return True
+
+    def allow_migrate(self, db, app_label, model_name=None, **hints):
+        return True
```
Comment (patrys) [inline saleor/core/db_routers.py:18]: Why implement this?

**Comment by dekoza** [comment]
OK, so basically I should only leave `allow_relation`, right?

File: saleor/core/db_routers.py (hunk 2 above)
Comment (dekoza) [reply saleor/core/db_routers.py]: Removed.

File: saleor/core/db_routers.py (hunk 1 above)
Comment (dekoza) [reply saleor/core/db_routers.py]: Removed.

Comment (dekoza) [reply saleor/core/db_routers.py]: Removed.

File: saleor/core/db_routers.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+class PrimaryReplicaRouter:
+    def allow_relation(self, obj1, obj2, **hints):
```
Comment (fowczarek) [inline saleor/core/db_routers.py:2]:
```suggestion
class PrimaryReplicaRouter:

    def db_for_write(self, model, **hints):
        """Write only to primary."""
        return settings.DATABASE_CONNECTION_DEFAULT_NAME

    def allow_relation(self, obj1, obj2, **hints):
```

**Review by fowczarek — APPROVED** [review]

File: saleor/settings.py [hunk 4]
```diff
@@ -89,6 +89,8 @@ def get_bool_from_env(name, default_value):
     # ),
 }
 
+DATABASE_ROUTERS = ["saleor.core.db_routers.PrimaryReplicaRouter"]
```
Comment (cmiacz) [inline saleor/settings.py:92]: We'll have to override this class in cloud builds, however methods used here does not conflict with the DB router for cloud.

Comment (cmiacz) [reply saleor/settings.py:92]: What Saleor versions will have this change?

**Review by cmiacz — APPROVED** [review]

File: saleor/settings.py (hunk 4 above)
Comment (dekoza) [reply saleor/settings.py:92]: Right now this goes only to main, so 3.8 onwards.

**Review by IKarbowiak — APPROVED** [review]

**Pull request merged**